src/desktop/tabletinput.cpp

```diff
@@ -18,73 +19,223 @@ namespace tabletinput {
 
 static const char *inputMode = "Qt tablet input";
 
-void init(QApplication *app, const QSettings &cfg)
-{
-#if defined(Q_OS_WIN) && defined(HAVE_KIS_TABLET)
-	bool useWindowsInk = false;
-	// Enable Windows Ink tablet event handler
-	// This was taken directly from Krita
-	if(cfg.value("settings/input/windowsink", true).toBool()) {
-		KisTabletSupportWin8 *penFilter = new KisTabletSupportWin8();
-		if(penFilter->init()) {
-			app->installNativeEventFilter(penFilter);
-			useWindowsInk = true;
-			inputMode = "KisTablet Windows Ink input";
-			qDebug("Using Win8 Pointer Input for tablet support");
+#ifdef Q_OS_WIN
 
-		} else {
-			qWarning("No Win8 Pointer Input available");
-			delete penFilter;
+static Mode currentMode = Mode::UNINITIALIZED;
+
+#	ifdef HAVE_KIS_TABLET
+
+class SpontaneousTabletEventFilter final : public QObject {
+public:
+	explicit SpontaneousTabletEventFilter(QObject *parent)
+		: QObject{parent}
+	{
+	}
+
+protected:
+	bool eventFilter(QObject *watched, QEvent *event) override final
+	{
+		switch(event->type()) {
+		case QEvent::TabletEnterProximity:
+		case QEvent::TabletLeaveProximity:
+		case QEvent::TabletMove:
+		case QEvent::TabletPress:
+		case QEvent::TabletRelease:
+			// KisTablet uses QApplication::sendEvent, which means its events
+			// are never set to be spontaneous. We use that to filter out Qt's
+			// own tablet events, which in turn are always spontaneous.
+			if(event->spontaneous()) {
```

Comment:
I’m not saying that anything needs to change if this works, but I am curious if there is any particular reason why the KisTablet events shouldn’t be just using their own registered custom event type and then all of the Qt tablet events are simply ignored when KisTablet is on? Using the same event types provides better isolation between the consumer and producer, but code is already doing stuff based on KisTablet ifdefs anyway, so it doesn’t seem like there’s much practical benefit but an increased risk of future breakage. Is this true?

Reply:
Parts of Qt react to tablet events and if the regular Qt tablet input is broken for your device, turning on KisTablet would only fix the canvas view and the tablet tester, while the rest of the UI would still receive broken events. Qt also does mouse event synthesis if you don't accept tablet events (I think that's how it works anyway), so we couldn't just switch over to reading the other kind of event in KisTablet mode, we'd still have to handle the Qt tablet events to explicitly throw them on the floor. So just blocking the Qt tablet events and producing our own seems more stable.

Reply:
OK. This is all just for future reference, since I bothered to look:

`Qt::AA_SynthesizeMouseForUnhandledTabletEvents` appears to control Qt’s automatic synthesis if the platform itself doesn’t already generate its own synthetic mouse events. Things I have noticed that it does with tablet events is (1) update the QGuiApplication modifier buttons state, (2) track what window is being targeted for events, (3) forward and/or ignore events in the same way as mouse events.

Reply:
Ah, that would explain why we get such different behavior on Windows and Linux, maybe one of those platforms generates synthetic events on its own. The tablet tester on Windows always shows both a tablet trail and a mouse trail, while Linux only shows a tablet trail (for me anyway.)